src/codegen/debug.rs

```diff
@@ -0,0 +1,518 @@
+use std::{cell::RefCell, collections::HashMap, ops::Range};
+
+use inkwell::{
+    context::ContextRef,
+    debug_info::{
+        AsDIScope, DIBasicType, DICompileUnit, DICompositeType, DIDerivedType, DIFlags,
+        DIFlagsConstants, DIType, DWARFEmissionKind, DebugInfoBuilder,
+    },
+    module::Module,
+    values::{BasicMetadataValueEnum, GlobalValue},
+};
+
+use crate::{
+    ast::SourceRange,
+    datalayout::Byte,
+    diagnostics::Diagnostic,
+    index::Index,
+    typesystem::{DataType, DataTypeInformation, StringEncoding, BOOL_SIZE, CHAR_TYPE, WCHAR_TYPE},
+    DebugLevel, OptimizationLevel,
+};
+
+#[derive(PartialEq, Eq)]
+#[allow(non_camel_case_types)]
+enum DebugEncoding {
+    // DW_ATE_address = 0x01,
+    DW_ATE_boolean = 0x02,
+    DW_ATE_float = 0x04,
+    DW_ATE_signed = 0x05,
+    DW_ATE_unsigned = 0x07,
+    DW_ATE_UTF = 0x10,
+}
+
+impl From<DWARFEmissionKind> for DebugLevel {
+    fn from(kind: DWARFEmissionKind) -> Self {
+        match kind {
+            DWARFEmissionKind::Full => DebugLevel::Full,
+            _ => DebugLevel::None,
+        }
+    }
+}
+
+impl From<DebugLevel> for DWARFEmissionKind {
+    fn from(level: DebugLevel) -> Self {
+        match level {
+            DebugLevel::Full | DebugLevel::VariablesOnly => DWARFEmissionKind::Full,
+            _ => DWARFEmissionKind::None,
+        }
+    }
+}
+
+pub trait Debug<'ink> {
+    fn register_debug_type<'idx>(
+        &self,
+        name: &str,
+        datatype: &'idx DataType,
+        index: &'idx Index,
+    ) -> Result<(), Diagnostic>;
+
+    fn create_global_variable(
+        &self,
+        name: &str,
+        type_name: &str,
+        global_variable: GlobalValue<'ink>,
+    ) -> Result<(), Diagnostic>;
+
+    fn finalize(&self) -> Result<(), Diagnostic>;
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum DebugType<'ink> {
+    Basic(DIBasicType<'ink>),
+    Struct(DICompositeType<'ink>),
+    Derived(DIDerivedType<'ink>),
+    Composite(DICompositeType<'ink>),
+}
+
+impl<'ink> From<DebugType<'ink>> for DIType<'ink> {
+    fn from(t: DebugType<'ink>) -> Self {
+        match t {
+            DebugType::Basic(t) => t.as_type(),
+            DebugType::Struct(t) => t.as_type(),
+            DebugType::Derived(t) => t.as_type(),
+            DebugType::Composite(t) => t.as_type(),
+        }
+    }
+}
+
+pub struct DebugObj<'ink> {
+    context: ContextRef<'ink>,
+    debug_info: DebugInfoBuilder<'ink>,
+    compile_unit: DICompileUnit<'ink>,
+    types: RefCell<HashMap<String, DebugType<'ink>>>,
+    global_metadata: RefCell<Vec<BasicMetadataValueEnum<'ink>>>,
+}
+
+/// Wraps a debug object in an enum based on its level
+pub enum DebugWrapper<'ink> {
+    None,
+    VariablesOnly(DebugObj<'ink>),
+    Full(DebugObj<'ink>),
+}
+
+pub fn new<'ink>(
```

Comment:
shouldnt this be in the context of DebugWrapper? its a pub method in this module